drivers/crc/crc_stm32.c

```diff
@@ -0,0 +1,232 @@
+			   uint32_t flags, crc_polynomial_t polynomial, const void *initial_value)
+{
+	/* Attempt to take mutex lock */
+	const struct crc_data *data = dev->data;
+	int mtx = crc_stm32_lock(dev, data->timeout);
+
+	if (mtx != 0) {
+		return mtx;
+	}
+
+	const struct crc_cfg *cfg = dev->config;
+
+	LL_CRC_SetInputDataReverseMode(cfg->base, (ctx->flags & CRC_FLAG_REVERSE_INPUT)
+							  ? LL_CRC_INDATA_REVERSE_WORD
+							  : LL_CRC_INDATA_REVERSE_NONE);
+	LL_CRC_SetOutputDataReverseMode(cfg->base, (ctx->flags & CRC_FLAG_REVERSE_OUTPUT)
+							   ? LL_CRC_OUTDATA_REVERSE_BIT
+							   : LL_CRC_INDATA_REVERSE_NONE);
+	LL_CRC_ResetCRCCalculationUnit(cfg->base);
+
+	switch (ctx->type) {
+	case CRC8_CCITT: {
+		uint8_t init_val = *(const uint8_t *)(ctx->initial_value);
+		uint8_t poly = (uint8_t)(ctx->polynomial & 0xFF);
+		LL_CRC_SetPolynomialSize(cfg->base, LL_CRC_POLYLENGTH_8B);
+		LL_CRC_SetPolynomialCoef(cfg->base, poly);
+		LL_CRC_SetInitialData(cfg->base, init_val);
+		break;
+	}
+	case CRC32_IEEE: {
+		uint32_t init_val = *(const uint32_t *)(ctx->initial_value);
+		uint32_t poly = (uint32_t)(ctx->polynomial & 0xFFFFFFFF);
+		LL_CRC_SetPolynomialSize(cfg->base, LL_CRC_POLYLENGTH_32B);
+		LL_CRC_SetPolynomialCoef(cfg->base, poly);
+		LL_CRC_SetInitialData(cfg->base, init_val);
+		break;
+	}
+	default:
+		return -ENOTSUP;
+	}
+	return 0;
+}
+
+static int crc_stm32_update(const struct device *dev, struct crc_ctx *ctx, const void *buffer,
+			    size_t bufsize)
+{
+	/* Check that the calling thread has access to the resource */
+	struct crc_data *data = (struct crc_data *)dev->data;
+	int mtx = crc_lock(dev, data->timeout);
+
+	if (mtx != 0) {
+		return mtx;
+	}
+
+	const struct crc_cfg *cfg = dev->config;
+	const uint8_t *num_data = buffer;
+
+	switch (ctx->type) {
+	case CRC8_CCITT: {
+		register uint32_t index = 0;
+
+		for (index = 0; index < bufsize; index++) {
+			LL_CRC_FeedData8(cfg->base, num_data[index]);
+		}
+		uint8_t result = LL_CRC_ReadData8(cfg->base);
+		(ctx->result) = (crc_result_t)result;
+		break;
+	}
+	case CRC32_IEEE: {
+		register uint32_t data = 0;
+		register uint32_t index = 0;
+
+		/* Compute the CRC of Data Buffer array*/
+		for (index = 0; index < (bufsize / 4); index++) {
+			data = (uint32_t)((num_data[4 * index + 3] << 24) |
+					  (num_data[4 * index + 2] << 16) |
+					  (num_data[4 * index + 1] << 8) | num_data[4 * index]);
+			LL_CRC_FeedData32(cfg->base, data);
+		}
+
+		/* Last bytes specific handling */
+		if ((bufsize & 3) != 0) {
+			if ((bufsize & 3) == 1) {
+				LL_CRC_FeedData8(cfg->base, num_data[4 * index]);
+			}
+			if ((bufsize & 3) == 2) {
+				LL_CRC_FeedData16(cfg->base,
+						  (uint16_t)((num_data[4 * index + 1] << 8) |
+							     num_data[4 * index]));
+			}
+			if ((bufsize & 3) == 3) {
+				LL_CRC_FeedData16(cfg->base,
+						  (uint16_t)((num_data[4 * index + 1] << 8) |
+							     num_data[4 * index]));
+				LL_CRC_FeedData8(cfg->base, num_data[4 * index + 2]);
+			}
```

Comment:
Not correct. This feeds data twice if `(bufsize & 3) == 3`.
Also, `sys_get_beXX` from `<zephyr/sys/byteorder.h>` should be used instead of manually unpacking.
```suggestion
			if ((bufsize & 3) == 3) {
			       LL_CRC_FeedData16(cfg->base, sys_get_be16(num_data + 4 * index));
			       LL_CRC_FeedData8(cfg->base, num_data[4 * index + 2]);
			} else if ((bufsize & 3) == 2) {
			       LL_CRC_FeedData16(cfg->base, sys_get_be16(num_data + 4 * index));
			} else { /* ((bufsize & 3) == 1) */
	                       LL_CRC_FeedData8(cfg->base, num_data[4 * index]);		
			}
```